Fixes #24256 - Optimize DB indices

tbrisker opened the pull request:
This removes some indices that aren't really needed since they are
already covered by parts of another index - both PostgreSQL and MySQL
know how to use partial indices. Additionally, a few polymorphic indices
were reversed in order so that the type column is indexed before the id
column. A couple of duplicate indices were also deleted.



<!---

Thank you for contributing to The Foreman, please read the
[following guide](https://www.theforeman.org/contribute.html), in short:

* [Create an issue](https://projects.theforeman.org/projects/foreman/issues)
* Reference the issue via `Fixes #1234` or `Refs #1234` in the commit message
* Prefer present-tense, imperative-style commit messages
* Extract all strings for i18n, see [Translating section in the guide]
(https://projects.theforeman.org/projects/foreman/wiki/Translating)
* Prepend `[WIP]` for work in progress from bots triggering actions
* Be patient, we will do our best to take a look as soon as we can
* Explain the purpose of the PR, attach screenshots if applicable
* Be nice and respectful

We are running bots that will poke you if you miss an item from the list :-)

--->

tbrisker committed:
Fixes #24256 - Optimize DB indices

This removes some indices that aren't really needed since they are
already covered by parts of another index - both PostgreSQL and MySQL
know how to use a subset of a multi-column index.
Additionally, a few polymorphic indices were reversed in order so
that the type column is indexed before the id column.
A couple of duplicate indices were also deleted.

tbrisker commented on db/migrate/20180715202514_optimize_indices.rb:
```diff
@@ -0,0 +1,45 @@
+    remove_index :notification_recipients, :user_id
+
+    remove_index :parameters, name: "index_parameters_on_domain_id_and_type" if index_exists?(:parameters, :type, name: "index_parameters_on_domain_id_and_type")
+    remove_index :parameters, name: "index_parameters_on_hostgroup_id_and_type" if index_exists?(:parameters, :type, name: "index_parameters_on_hostgroup_id_and_type")
```
These two may be leftovers from `db/migrate/20100525094200_simplify_parameters.rb` migration on certain dbs

tbrisker commented on db/migrate/20180715202514_optimize_indices.rb:
```diff
@@ -0,0 +1,45 @@
+
+    remove_index :hosts, :type
+
+    remove_index :lookup_values, name: "index_lookup_values_on_priority" if index_exists?(:lookup_values, :match, name: "index_lookup_values_on_priority")
```
This may be a leftover from when priority was renamed to match

lzap commented on db/migrate/20100525094200_simplify_parameters.rb:
```diff
@@ -1,11 +1,11 @@
-    add_index     :parameters, [:reference_id, :type] if index_exists? :parameters, :reference_id
+    add_index     :parameters, [:reference_id, :type]
 
     Parameter.reset_column_information
```
This appears like change in already released migration from 2010. We must be absolutely sure there are not side effects. And I think there can be one, I suggest to create a copy of these statements in a new migration (we can keep those changed but the copy will make sure the same is executed for upgraded users).

tbrisker replied:
tbh i'm not sure we need this index at all, we already have index on `["type", "reference_id", "name"]` and I don't see any sense in looking for a parameter with just reference id and no type.

tbrisker replied:
I'll just revert the change to the migration and clean up any leftover indices in my migration.

lzap commented:
Can you elaborate what you mean by "PostgreSQL and MySQL know how to use partial indices"? I know what partial index is and how to make use of it in Rails, but I don't see us doing that in our current codebase:

    # ag add_index | grep where

I am failing to see the connection between that and the PR changes.

lzap commented on db/migrate/20180715202514_optimize_indices.rb:
```diff
@@ -0,0 +1,45 @@
+    remove_index :bookmarks, [:owner_id, :owner_type]
+    add_index :bookmarks, [:owner_type, :owner_id]
+
+    remove_index :fact_names, :ancestry
```
`scope :root_only, -> { with_roots.where(:fact_names => {:ancestry => nil}) }`

Granted the "name" can follow, but there is just one use which does not use it.

Partial index should help here, but partitioning would be the best. Not supported in vanilla Rails tho, so I'd suggest:

    remove_index :fact_names, :ancestry, where: {ancestry: nil}

tbrisker replied:
There is index `["ancestry", "name"]` that can be used for this as well

lzap commented on db/migrate/20180715202514_optimize_indices.rb:
```diff
@@ -0,0 +1,45 @@
+
+    remove_index :fact_names, :ancestry
+
+    remove_index :fact_values, :fact_name_id
```
Couple of finds:

```
changes = FactValue.where(:fact_name_id => fact_name_id).group(:value).pluck(:value)

FactValue.select('fact_name_id, value').group(:fact_name_id, :value).where(:fact_name_i => fact_name_id).includes(:fact_name).map

deleted_values = FactValue.unscoped.where(:fact_name_id => fact_names).delete_all
```

The last one is fact cleaner which is performance-sensitive part of Foreman.

tbrisker replied:
There is another an index on `["fact_name_id", "host_id"]` which can also be used.

lzap commented:
Oh now I understand. I think you used *partial index* term instead *multicolumn index* term, at least this is how it called for both PostgreSQL and MySQL:

* https://www.postgresql.org/docs/current/static/indexes-partial.html
* https://www.postgresql.org/docs/current/static/indexes-multicolumn.html

That's why I was confused. Frankly, I had no idea that MySQL ever supported that. It looks like 5.5+ supports that. This is what I was able to find:

* PostgreSQL 7.2+ https://www.postgresql.org/docs/7.2/static/indexes-multicolumn.html
* MySQL 5.5+ https://dev.mysql.com/doc/refman/5.5/en/multiple-column-indexes.html (also found some old mirrors with 5.1 documentation copy)

Looks pretty safe then. Anyway, there is some gem which can help us to optimize further: https://github.com/gregnavis/active_record_doctor

```
[lzap@box foreman]$ be rake active_record_doctor:extraneous_indexes
The following indexes are extraneous and can be removed:
  index_fact_names_on_ancestry (can be handled by index_fact_names_on_ancestry_and_name)
  index_taxable_taxonomies_on_taxable_id_and_taxable_type (can be handled by taxable_index)
  index_user_roles_on_owner_id (can be handled by index_user_roles_on_owner_id_and_owner_type)
  associated_index (can be handled by )
  auditable_parent_index (can be handled by )
  index_hosts_on_type (can be handled by index_hosts_on_type_and_location_id, index_hosts_on_type_and_organization_id_and_location_id)
  index_reports_on_type (can be handled by index_reports_on_type_and_host_id)
  index_reports_on_reported_at (can be handled by index_reports_on_reported_at_and_host_id_and_type)
  index_reports_on_host_id (can be handled by index_reports_on_host_id_and_type_and_id)
  index_reports_on_reported_at_and_host_id (can be handled by index_reports_on_reported_at_and_host_id_and_type)
  index_by_type (can be handled by index_by_type_and_id)
  index_notification_recipients_on_user_id (can be handled by index_notification_recipients_on_user_id_and_notification_id)
  index_fact_values_on_fact_name_id (can be handled by index_fact_values_on_fact_name_id_and_host_id)
  index_foreman_tasks_locks_on_resource_type (can be handled by index_foreman_tasks_locks_on_resource_type_and_resource_id)
  index_foreman_tasks_locks_on_name (can be handled by index_foreman_tasks_locks_name_resource_type_resource_id)
  index_foreman_tasks_tasks_on_id (can be handled by index_foreman_tasks_id_state)
  index_foreman_tasks_tasks_on_type (can be handled by index_foreman_tasks_tasks_on_type_and_label)
  index_parameters_on_type (can be handled by index_parameters_on_type_and_reference_id_and_name)
  index_parameters_on_domain_id_and_type (can be handled by index_parameters_on_type_and_reference_id_and_name)
  index_parameters_on_hostgroup_id_and_type (can be handled by index_parameters_on_type_and_reference_id_and_name)
  index_audits_on_id (is a primary key of audits)
  index_foreman_tasks_tasks_on_id (is a primary key of foreman_tasks_tasks)
```

This confirms most if not all your changes, I haven't checked one by one. Would you mind reviewing that and only pointing on those which were not found by the tool.

The tool looks pretty useful:

```
[lzap@box foreman]$ be rake active_record_doctor:unindexed_foreign_keys
architectures_operatingsystems architecture_id operatingsystem_id
compute_resources http_proxy_id
config_group_classes config_group_id puppetclass_id
discovery_rules hostgroup_id
domains dns_id
environment_classes puppetclass_lookup_key_id
external_usergroups auth_source_id
features_smart_proxies feature_id smart_proxy_id
filters role_id
foreman_tasks_recurring_logics triggering_id
foreman_tasks_remote_tasks remote_task_id step_id
foreman_tasks_task_group_members task_group_id
host_classes host_id puppetclass_id
host_config_groups config_group_id host_id
hostgroups architecture_id compute_resource_id domain_id environment_id medium_id operatingsystem_id ptable_id puppet_ca_proxy_id puppet_proxy_id realm_id salt_environment_id salt_proxy_id subnet6_id subnet_id
hosts compute_resource_id discovery_rule_id image_id location_id model_id organization_id owner_id ptable_id puppet_ca_proxy_id puppet_proxy_id realm_id salt_environment_id salt_proxy_id
images architecture_id compute_resource_id operatingsystem_id
key_pairs compute_resource_id
locations_organizations location_id organization_id
lookup_values lookup_key_id
media_operatingsystems medium_id operatingsystem_id
nics domain_id subnet6_id subnet_id
operatingsystems_provisioning_templates operatingsystem_id provisioning_template_id
operatingsystems_ptables operatingsystem_id ptable_id
operatingsystems_puppetclasses operatingsystem_id puppetclass_id
os_default_templates operatingsystem_id provisioning_template_id template_kind_id
realms realm_proxy_id
roles cloned_from_id
salt_module_environments salt_environment_id salt_module_id
subnet_domains domain_id
subnets dhcp_id discovery_id dns_id template_id tftp_id
template_combinations environment_id hostgroup_id provisioning_template_id
templates template_kind_id
user_mail_notifications mail_notification_id user_id
user_roles role_id
usergroup_members member_id usergroup_id
users auth_source_id default_location_id default_organization_id
```

That also looks promising, it's hard to believe all of these are missing but quick look into schema.rb confirmed. This creates unfortunate situation - you wanted to drop some and this tells us to add more. We can probably do a separate PR.

tbrisker commented:
@lzap - correct, i was referring to using a subset of a multi-column index, sorry about the confusion. I'll look at the results from the tool to make sure I didn't miss anything. Also I'm not sure regarding adding the indices to all fk's but in any case that should be a separate pr.

tbrisker commented:
@lzap - I've updated this PR, removed the changes to the old migration and added comments inline to explain each change. some of the indices the tool indicated were from task plugin (i guess we need a similar pr there as well). There were also a few it missed or were irrelevant once I changed the order of the polymorphic indices. There was one that was missing from my initial pr which I added (audit index on id).

ekohl commented:
@tbrisker I do know that MySQL can be picky about multi-column indices (from the page @lzap linked):

> If the table has a multiple-column index, any leftmost prefix of the index can be used by the optimizer to look up rows. For example, if you have a three-column index on (col1, col2, col3), you have indexed search capabilities on (col1), (col1, col2), and (col1, col2, col3).
>
> If an index exists on (col1, col2, col3), only the first two queries use the index. The third and fourth queries do involve indexed columns, but do not use an index to perform lookups because (col2) and (col2, col3) are not leftmost prefixes of (col1, col2, col3).

I haven't looked at the actual indexes, but this is an easy mistake to make.

tbrisker commented:
@ekohl all of the indices removed are left-most prefix of another index. I believe PG also works the same way - only left-prefix subset can be used.

lzap commented:
Sorry for the delay, since the `active_record_doctor` gem proved to be useful, shall we add it as a development dependency? That won't hurt. It offers integration with Jenkins too, that would be great - preventing from the beginning.

tbrisker commented:
@lzap i'm fine with that but lets do that in a separate issue

lzap reviewed (approved):
Looks great, according to active_record_doctor no extra indicies found, it migrates fine. However this can have performance side effect I am unable to test with my small database. Let's see. I was wondering if we can try to cherry pick it into 1.19 so we can clearly recognize RDBM regressions because three months later we will definitely forget about this PR.
